pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -415,6 +424,128 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 	return reconcile.Result{}, nil
 }
 
+// updateInventory ensures that the inventory of the cluster pool and the related ClusterDeploymentCustomizations are up to date.
+// This includes validating resources existance and updating references when needed.
+func (r *ReconcileClusterPool) updateInventory(clp *hivev1.ClusterPool, cds []*hivev1.ClusterDeployment, status hivev1.LastApplyStatusType, logger log.FieldLogger) {
+	// no need to update Inventory if it doesn't exist
+	if clp.Spec.Inventory == nil {
+		return
+	}
+	// InventoryValid condition and reason
+	reason := hivev1.InventoryReasonValid
+	switch status {
+	case hivev1.LastApplyBrokenCloud:
+		reason = hivev1.InventoryReasonBrokenByCloud
+	case hivev1.LastApplyBrokenSyntax:
+		reason = hivev1.InvenotryReasonBrokenBySyntax
+	}
+
+	// Helper functions to get and update ClusterDeploymentCustomization, and update inventory message if missing
+	getCDC := func(name string) *hivev1.ClusterDeploymentCustomization {
+		cdc := &hivev1.ClusterDeploymentCustomization{}
+		if err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: clp.Namespace, Name: name}, cdc); err != nil {
+			if apierrors.IsNotFound(err) {
+				reason := hivev1.InventoryReasonMissing
+				r.updateInventoryValidMessage(clp, cdc.Name, reason, false, logger)
+			}
+			log.WithError(err).Warn("error reading customization")
+			return nil
+		}
+		return cdc
+	}
+
+	updateCDC := func(cdc *hivev1.ClusterDeploymentCustomization, status bool, msg string) {
+		if err := r.setCustomizationAvailabilityCondition(cdc, status, msg, logger); err != nil {
+			log.WithError(err).Warn("failed to update customization status")
+		}
+	}
+
+	// First update the inventory and ClusterDeploymentCustomizations related to changed ClusterDeployments
+	var active_cdc []string
+	for _, cd := range cds {
+		active_cdc = append(active_cdc, cd.Spec.ClusterPoolRef.CustomizationRef.Name)
```

Comment:
This might not be set for all CDs, e.g. if we've recently added the inventory to the pool and are still rotating the old ones out.
```suggestion
		if cd.Spec.ClusterPoolRef.CustomizationRef == nil {
			continue
		}
		active_cdc = append(active_cdc, cd.Spec.ClusterPoolRef.CustomizationRef.Name)
```